feat: Brillig pointer codegen and execution

vezenovm opened the pull request:
Resolves https://github.com/noir-lang/noir/issues/3907

This PR builds upon https://github.com/AztecProtocol/aztec-packages/pull/5709. 

These changes do not yet include a Brillig stdlib and removal of the `Brillig` opcode itself. The generated stdlib Brillig (such as for quotient) is not created in the same manner as other Brillig calls which are generated during SSA. I have decided to leave this for another follow-up where we can actually remove `Brillig`.

vezenovm committed:
initial BrilligPointer opcode without any implementation

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
initial codegen work for brillig pointer and execution in the acvm

vezenovm committed:
update circuit serialization and instance deployer constant

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
revert commented hash check

vezenovm committed:
update todo comment

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
Merge branch 'mv/brillig-pointer' into mv/codegen-brillig-pointer

vezenovm committed:
handle brillig pointer in acvm js and cargo fmt

vezenovm committed:
cleanup BrilligSolver new methods

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
some cleanup

vezenovm committed:
cargo fmt

TomAFrench committed:
Apply suggestions from code review

vezenovm committed:
use Vec inside of map for generated brilligs

vezenovm committed:
cargo fmt

vezenovm committed:
fix to use generated index when setting brillig pointer

vezenovm committed:
cargo mft

vezenovm committed:
update instance deployer address

vezenovm committed:
some improvements, store GeneratedBrilligs during SSA as a Vec

vezenovm committed:
make both brillig func info a mutable reference on an AcirContext

TomAFrench committed:
Update noir/noir-repo/tooling/nargo_cli/src/cli/info_cmd.rs

vezenovm committed:
change serde to BrilligCall and no inner struct

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
undo comment in serde test

vezenovm committed:
Merge remote-tracking branch 'origin/mv/brillig-pointer' into mv/brillig-pointer

vezenovm committed:
Update noir/noir-repo/acvm-repo/acir/src/circuit/opcodes.rs

Co-authored-by: Tom French <15848336+TomAFrench@users.noreply.github.com>

vezenovm committed:
brillig call display

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
fix mem op serde

vezenovm committed:
Merge remote-tracking branch 'origin/mv/brillig-pointer' into mv/brillig-pointer

vezenovm committed:
merge conflcits w/ mv/brillig-pointer

TomAFrench committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
working again after merge conflicts

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
move to shraed context in acir gen

vezenovm committed:
cleanup

vezenovm committed:
cleanup comment in test

vezenovm committed:
Merge branch 'mv/brillig-pointer' into mv/codegen-brillig-pointer

vezenovm committed:
Merge branch 'master' into mv/brillig-pointer

vezenovm committed:
Merge branch 'mv/brillig-pointer' into mv/codegen-brillig-pointer

vezenovm committed:
update serde

sirasistant commented:
I think this needs a master merge in to cleanup the diff :)

vezenovm committed:
merge conflicts w/ master

vezenovm committed:
update instance delployer address

vezenovm committed:
Merge branch 'master' into mv/codegen-brillig-pointer

jfecher commented on noir/noir-repo/test_programs/execution_success/brillig_acir_loop/src/main.nr:
```diff
@@ -0,0 +1,14 @@
+fn main(x: Field, y: pub Field) {
+    for _ in 0..4 {
+        foo(x, y);
+        check_values(x, y);
+    }
+}
```
What does this test check for? That the loop doesn't generate extra copies of the brillig byte code?

If so I wonder if this could be a unit test somehow so we could attempt to check for that directly since this only tests that it executes correctly.

vezenovm replied:
Yeah I was using it to test my codegen, but you are correct that in `execution_success` it doesn't make much sense so I am going to move it to a unit test in ACIR gen.

vezenovm replied:
Made a unit test and deleted this test as it is unnecessary

vezenovm committed:
Update noir/noir-repo/acvm-repo/acvm/src/pwg/mod.rs

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm commented on noir/noir-repo/acvm-repo/acvm/src/pwg/mod.rs:
```diff
@@ -165,10 +165,22 @@ pub struct ACVM<'a, B: BlackBoxFunctionSolver> {
+    // whether the caller has resolved the bytecode of a Brillig pointer
+    // brillig_call_counter: usize,
+    // Represents the bytecode pointer
+    // brillig_bytecodes: Vec<
```
Oops yups old comments I missed

vezenovm committed:
Update noir/noir-repo/compiler/noirc_evaluator/src/ssa.rs

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm committed:
Update noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm commented on noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs:
```diff
@@ -1553,6 +1554,105 @@ impl AcirContext {
+        unsafe_return_values: bool,
+        brillig_function_index: u32,
+    ) -> Result<Vec<AcirValue>, RuntimeError> {
+        let b_inputs = try_vecmap(inputs, |i| -> Result<_, InternalError> {
```
This is from the old Brillig method, but I will change all of them I think it is clearer to have the prefix.

vezenovm committed:
Update noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm committed:
Update noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm committed:
Update noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm committed:
Update noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm committed:
Update noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm commented on noir/noir-repo/compiler/noirc_evaluator/src/ssa/ssa_gen/program.rs:
```diff
@@ -27,7 +27,29 @@ impl Ssa {
-        let id_to_index = btree_map(functions.iter().enumerate(), |(i, (id, _))| (*id, i as u32));
+        let mut acir_index = 0;
+        let mut brillig_index = 0;
+        let id_to_index = btree_map(functions.iter().enumerate(), |(_, (id, func))| {
```
This was used to get the final generated index of the ACIR entry point function. I had also wanted to avoid a shared context in ACIR gen, but now with Brillig it is unavoidable. I'm going to move to a new strategy without this map at all and just using a func id -> generated index map in the Acir gen SharedContext.

vezenovm replied:
Actually it does look easier to keep it just for ACIR functions as we do not codegen the ACIR upon making a call but separately for each SSA function. I will update with a comment.

vezenovm commented on noir/noir-repo/cspell.json:
```diff
@@ -27,6 +27,7 @@
         "boilerplates",
         "bridgekeeper",
         "brillig",
+        "brilligs",
```
:satisfied:

vezenovm committed:
Update noir/noir-repo/cspell.json

Co-authored-by: jfecher <jfecher11@gmail.com>

vezenovm committed:
Merge branch 'master' into mv/codegen-brillig-pointer

vezenovm committed:
rename generated_brilligs

vezenovm committed:
back to loop for brillig array input

vezenovm commented on noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs:
```diff
@@ -1553,6 +1554,105 @@ impl AcirContext {
+            match i {
+                AcirValue::Var(var, _) => Ok(BrilligInputs::Single(self.var_to_expression(var)?)),
+                AcirValue::Array(vars) => {
+                    let mut var_expressions: Vec<Expression> = Vec::new();
+                    for var in vars {
+                        self.brillig_array_input(&mut var_expressions, var)?;
+                    }
```
I changed this back to a loop as `brillig_array_input` takes in a mutable reference that we build up

vezenovm committed:
missed name

vezenovm committed:
Merge branch 'master' into mv/codegen-brillig-pointer

vezenovm committed:
label the entry point indices

vezenovm committed:
add unit test for brilligcall  gen and delete brillig_acir_loop

vezenovm committed:
merge conflicts

vezenovm commented:
Added a unit test here [477840d](https://github.com/AztecProtocol/aztec-packages/pull/5737/commits/477840dd339a47a650e9748f661c6904dc3f0bd0) and switched to using function ids for generating the brillig call pointers during ACIR gen.

vezenovm commented on noir/noir-repo/compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:
```diff
@@ -39,9 +40,48 @@ use im::Vector;
+    /// There can be Brillig functions specified in SSA which do not act as
+    /// entry points in ACIR (e.g. only called by other Brillig functions)
+    /// This mapping is necessary to use the correct function pointer for a Brillig call.
+    brillig_generated_func_pointers: BTreeMap<u32, u32>,
```
Done

vezenovm committed:
Merge branch 'master' into mv/codegen-brillig-pointer

vezenovm committed:
update constants
